ambari-server/src/main/java/org/apache/ambari/server/checks/UpgradeCheckRegistry.java

```diff
@@ -71,90 +80,59 @@ public void register(UpgradeCheck upgradeCheck) {
-    for (String prerequisiteCheck : prerequisiteChecks) {
-      if (!isRegistered(prerequisiteCheck)) {
-        missingChecks.add(prerequisiteCheck);
-      }
-    }
-
-    List<UpgradeCheck> checks = new ArrayList<>(missingChecks.size());
-    if (missingChecks.isEmpty()) {
-      return checks;
-    }
-
-    List<URL> urls = new ArrayList<>();
-    for (ServiceInfo service : services.values()) {
-      File dir = service.getChecksFolder();
-      File[] jars = dir.listFiles(new FilenameFilter() {
-        @Override
-        public boolean accept(File dir, String name) {
-          return name.endsWith(".jar");
-        }
-      });
-      for (File jar : jars) {
-        try {
-          URL url = jar.toURI().toURL();
-          urls.add(url);
-          LOG.debug("Adding service check jar to classpath: {}", url);
-        }
-        catch (Exception e) {
-          LOG.error("Failed to add service check jar to classpath: {}", jar.getAbsolutePath(), e);
-        }
-      }
-    }
-
-    ClassLoader classLoader = new URLClassLoader(urls.toArray(new URL[urls.size()]), ClassUtils.getDefaultClassLoader());
-    for (String prerequisiteCheck : missingChecks) {
-      Class<?> clazz = null;
-      try {
-        clazz = ClassUtils.resolveClassName(prerequisiteCheck, classLoader);
-      }
-      catch (IllegalArgumentException illegalArgumentException) {
-        LOG.error("Unable to find upgrade check {}", prerequisiteCheck, illegalArgumentException);
-      }
-      try {
-        if (clazz != null) {
-          UpgradeCheck upgradeCheck = (UpgradeCheck) clazz.newInstance();
-          checks.add(upgradeCheck);
-        }
-      } catch (Exception exception) {
-        LOG.error("Unable to create upgrade check {}", prerequisiteCheck, exception);
-      }
-    }
-    return checks;
-  }
-
-  private boolean isRegistered(String prerequisiteCheck) {
-    for (UpgradeCheck descriptor: m_upgradeChecks){
-      if (prerequisiteCheck.equals(descriptor.getClass().getName())){
-        return true;
-      }
-    }
-    return false;
+    return new ArrayList<>(m_builtInChecks);
   }
 
   /**
    * Gets an ordered and filtered list of the upgrade checks.
    * @param upgradePack Upgrade pack object with the list of required checks to be included
    * @return
    */
-  public List<UpgradeCheck> getFilteredUpgradeChecks(UpgradePack upgradePack){
-    List<String> prerequisiteChecks = upgradePack.getPrerequisiteChecks();
-    List<UpgradeCheck> resultCheckDescriptor = new ArrayList<>();
-    for (UpgradeCheck descriptor: m_upgradeChecks){
-      if (isRequired(descriptor, upgradePack.getType())) {
-        resultCheckDescriptor.add(descriptor);
-      } else if (prerequisiteChecks.contains(descriptor.getClass().getName())){
-        resultCheckDescriptor.add(descriptor);
+  @SuppressWarnings("unchecked")
+  public List<UpgradeCheck> getFilteredUpgradeChecks(UpgradePack upgradePack) throws AmbariException {
```

Comment:
We have the ability to parameterize some classes per upgrade pack.  Do we need to consider the case where an UP defines an Ambari class with properties?  Should open a jira if you think this is something we need to allow.

Reply:
Are you talking about something like:

```
 <check>org.apache.ambari.server.checks.ServicePresenceCheck</check>
...
<check-properties name="org.apache.ambari.server.checks.ServicePresenceCheck">
  <property name="no-upgrade-support-service-names">Atlas</property>
</check-properties>
```

I see what you're saying. So right now, here's what happens:
- The Ambari Framework will fine this class and load it as a "built-in". There's no check for `UpgradeCheckInfo#required`, so it just registers it outright.
- When an `UpgradePack` which has this check is loaded, it will load it as a plugin-check for that specific `UpgradePack`
- When the pre-checks run, this class will run twice; once with no properties and once with the properties from the stack.

I think you're right - we should probably only load `required` upgrade checks as built-in. Thoughts?